packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart

```diff
@@ -131,20 +147,136 @@ class FuchsiaDevice extends Device {
+    return parseFuchsiaDartPortOutput(lsOutput);
+  }
+
+  /// Run `command` on the Fuchsia device shell.
+  Future<String> shell(String command) async {
+    final RunResult result = await runAsync(<String>[
+      'ssh', '-F', fuchsiaSdk.sshConfig.absolute.path, id, command]);
+    if (result.exitCode != 0) {
+      throwToolExit('Command failed: $command\nstdout: ${result.stdout}\nstderr: ${result.stderr}');
+      return null;
+    }
+    return result.stdout;
+  }
+
+  /// Finds the first port running a VM matching `isolateName` from the
+  /// provided set of `ports`.
+  ///
+  /// TODO(jonahwilliams): replacing this with the hub will require an update
+  /// to the flutter_runner.
+  Future<int> findIsolatePort(String isolateName, List<int> ports) async {
+    for (int port in ports) {
+      try {
+        // TODO(awdavies): The square-bracket enclosure for using the IPv6 loopback
+        // didn't appear to work, but when assigning to the IPv4 loopback device,
+        // netstat shows that the local port is actually being used on the IPv6
+        // loopback (::1). While this can be used for forwarding to the destination
+        // IPv6 interface, it cannot be used to connect to a websocket.
+        final Uri uri = Uri.parse('http://[$_ipv6Loopback]:$port');
+        final VMService vmService = await VMService.connect(uri);
+        await vmService.getVM();
+        await vmService.refreshViews();
+        for (FlutterView flutterView in vmService.vm.views) {
+          if (flutterView.uiIsolate == null) {
+            continue;
+          }
+          final Uri address = flutterView.owner.vmService.httpAddress;
+          if (flutterView.uiIsolate.name.contains(isolateName)) {
+            return address.port;
+          }
+        }
+      } on SocketException catch (err) {
+        printTrace('Failed to connect to $port: $err');
+      }
+    }
+    throwToolExit('No ports found running $isolateName');
+    return null;
+  }
+}
+
+class _FuchsiaPortForwarder extends DevicePortForwarder {
+  _FuchsiaPortForwarder(this.device);
+
+  final FuchsiaDevice device;
+
+  @override
+  Future<int> forward(int devicePort, {int hostPort}) async {
+    hostPort ??= 0;
+    // Note: the provided command works around a bug in -N, but the solution is flaky.
+    final List<String> command = <String>[
+      'ssh', '-6', '-F', fuchsiaSdk.sshConfig.absolute.path, '-nNT', '-vvv', '-f',
+      '-L', '$hostPort:$_ipv4Loopback:$devicePort', device.id, 'true'
+    ];
+    final ProcessResult result = await processManager.run(command);
+    if (result.exitCode != 0) {
+      throwToolExit('Failed to forward port:$devicePort');
+    }
+    _forwardedPorts.add(ForwardedPort(hostPort, devicePort));
+    return hostPort;
+  }
+
+  @override
+  List<ForwardedPort> get forwardedPorts => _forwardedPorts;
+  final List<ForwardedPort> _forwardedPorts = <ForwardedPort>[];
+
+  @override
+  Future<void> unforward(ForwardedPort forwardedPort) async {
```

Comment:
This may not actually cause the process started in forward() above to end. It may need to be manually killed. See: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/fuchsia_reload.dart#L486. This is important to avoid colliding with subsequent attempts to set up a tunnel.

Reply:
Followed the example in fuchsia_attach - I close the ServerSocket and kill the original process